src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs

```diff
@@ -2133,6 +2133,1855 @@ struct S
+                );
+        }
+
+        [Fact]
+        public void ListPattern_InParametersInIndexAndRangeIndexers()
+        {
+            var source = @"
+new C().M();
+
+public class C
+{
+    public int Length => 2;
+    public string this[in System.Index i] => ""item value"";
+    public string this[in System.Range r] => ""rest value"";
+
+    public void M()
+    {
+        if (this is [ var item, ..var rest ])
+        {
+            System.Console.Write((item, rest));
+        }
+    }
+
+    void M2()
+    {
+        _ = this[^1];
+        _ = this[1..^1];
+    }
+}
+";
+            var compilation = CreateCompilation(new[] { source, TestSources.Index, TestSources.Range });
+            compilation.VerifyEmitDiagnostics();
+            var verifier = CompileAndVerify(compilation, expectedOutput: "(item value, rest value)");
+
+            verifier.VerifyIL("C.M", @"
+{
+  // Code size       79 (0x4f)
+  .maxstack  4
+  .locals init (string V_0, //item
+                string V_1, //rest
+                C V_2,
+                System.Index V_3,
+                System.Range V_4)
+  IL_0000:  ldarg.0
+  IL_0001:  stloc.2
+  IL_0002:  ldloc.2
+  IL_0003:  brfalse.s  IL_004e
+  IL_0005:  ldloc.2
+  IL_0006:  callvirt   ""int C.Length.get""
+  IL_000b:  ldc.i4.1
+  IL_000c:  blt.s      IL_004e
+  IL_000e:  ldloc.2
+  IL_000f:  ldc.i4.0
+  IL_0010:  ldc.i4.0
+  IL_0011:  newobj     ""System.Index..ctor(int, bool)""
+  IL_0016:  stloc.3
+  IL_0017:  ldloca.s   V_3
+  IL_0019:  callvirt   ""string C.this[in System.Index].get""
+  IL_001e:  stloc.0
+  IL_001f:  ldloc.2
+  IL_0020:  ldc.i4.1
+  IL_0021:  ldc.i4.0
+  IL_0022:  newobj     ""System.Index..ctor(int, bool)""
+  IL_0027:  ldc.i4.0
+  IL_0028:  ldc.i4.1
+  IL_0029:  newobj     ""System.Index..ctor(int, bool)""
+  IL_002e:  newobj     ""System.Range..ctor(System.Index, System.Index)""
+  IL_0033:  stloc.s    V_4
+  IL_0035:  ldloca.s   V_4
+  IL_0037:  callvirt   ""string C.this[in System.Range].get""
+  IL_003c:  stloc.1
+  IL_003d:  ldloc.0
+  IL_003e:  ldloc.1
+  IL_003f:  newobj     ""System.ValueTuple<string, string>..ctor(string, string)""
+  IL_0044:  box        ""System.ValueTuple<string, string>""
+  IL_0049:  call       ""void System.Console.Write(object)""
+  IL_004e:  ret
+}
+");
+        }
+
+        [Fact]
+        public void ListPattern_ImplicitlyConvertibleFromIndexAndRange()
+        {
+            var source = @"
+new C().M();
+
+public class MyIndex
+{
+    public static implicit operator MyIndex(System.Index i) => new MyIndex();
+}
+
+public class MyRange
+{
+    public static implicit operator MyRange(System.Range i) => new MyRange();
+}
+
+public class C
+{
+    public int Length => 2;
+    public string this[MyIndex i] => ""item value"";
+    public string this[MyRange r] => ""rest value"";
+
+    public void M()
+    {
+        if (this is [ var item, ..var rest ])
```

Comment:
Is this how range/index work in general?

Reply:
Yes, this is illustrated by `M2()` below (line 3509, only there for reference)